vowpalwabbit/learner.cc

```diff
@@ -36,207 +30,251 @@ const char* to_string(prediction_type_t prediction_type)
 
 namespace LEARNER
 {
-void dispatch_end_pass(vw& all, example& ec)
-{
-  all.current_pass++;
-  all.l->end_pass();
-  VW::finish_example(all, ec);
-}
-
-void save(vw& all, example* ec)
-{
-  // save state command
-  string final_regressor_name = all.final_regressor_name;
-
-  if ((ec->tag).size() >= 6 && (ec->tag)[4] == '_')
-    final_regressor_name = string(ec->tag.begin() + 5, (ec->tag).size() - 5);
-
-  if (!all.quiet)
-    all.trace_message << "saving regressor to " << final_regressor_name << endl;
-  save_predictor(all, final_regressor_name, 0);
-
-  VW::finish_example(all, *ec);
-}
-
-bool inline is_save_cmd(example* ec)
-{
-  return (ec->tag.size() >= 4) && (0 == strncmp((const char*)ec->tag.begin(), "save", 4));
-}
+  void learn_ex(example& ec, vw& all)
+  {
+    all.learn(ec);
+    as_singleline(all.l)->finish_example(all, ec);
+  }
+  
+  void learn_multi_ex(multi_ex& ec_seq, vw& all)
+  {
+    all.learn(ec_seq);
+    as_multiline(all.l)->finish_example(all, ec_seq);
+  }
+  
+  void end_pass(example& ec, vw& all)
+  {
+    all.current_pass++;
+    all.l->end_pass();
+    
+    VW::finish_example(all, ec);
+  }
+  
+  void save(example& ec, vw& all)
+  {
+    // save state command
+    string final_regressor_name = all.final_regressor_name;
 
-void process_example(vw& all, example* ec)
-{
-  if (ec->indices.size() > 1)  // 1+ nonconstant feature. (most common case first)
-    dispatch_example(all, *ec);
-  else if (ec->end_pass)
-    dispatch_end_pass(all, *ec);
-  else if (is_save_cmd(ec))
-    save(all, ec);
-  else
-    dispatch_example(all, *ec);
-}
+    if ((ec.tag).size() >= 6 && (ec.tag)[4] == '_')
+      final_regressor_name = string(ec.tag.begin() + 5, (ec.tag).size() - 5);
 
-void process_multi_ex(vw& all, multi_ex& ec_seq)
-{
-  all.learn(ec_seq);
-  as_multiline(all.l)->finish_example(all, ec_seq);
-}
+    if (!all.quiet)
+      all.trace_message << "saving regressor to " << final_regressor_name << endl;
+    save_predictor(all, final_regressor_name, 0);
 
-/* example headers have the word "shared" */
-bool ec_is_example_header(example& ec)
-{
-  v_array<CB::cb_class> costs = ec.l.cb.costs;
-  if (costs.size() != 1)
+    VW::finish_example(all, ec);
+  }
+  
+  /* example headers have the word "shared" */
+  bool ec_is_example_header(example& ec)
```

Comment:
I don't think this is going to work once CCB is in. But I have been running and testing with CCB for some time with an equivalent code path...

Reply:
not sure if I get it - this function is not changed at all and CCB PR does not have any changes in learner.cc, so it should have exactly the same version of code for it.